### PR TITLE
feat: add config sync

### LINKED ISSUE #286: [FEATURE]跨浏览器同步功能
### Before you begin

- [x] I have searched existing issues to avoid duplicates
- [ ] I am willing to contribute a PR for this feature

### Is your feature request related to a problem? Please describe.

跨设备使用比较麻烦，需要对每一台电脑的每一个浏览器的插件都进行重复配置操作；一旦某项设定需要变更，例如OpenAI的Key，则需要打开所有电脑的所有浏览器进行配置，使用负担较重

### Describe the solution you'd like

**建议可向两个方向进行优化：**

1. 支持WebDAV同步，可自定义配置WebDAV选项，包括网址路径、用户名、密码。
2. 支持类似于暴力猴插件的Google、OneDrive那样的授权同步。

**后期预期效果：**

1. 支持自动备份配置数据：插件可在配置变更时主动备份配置数据；
2. 支持自动还原配置数据：发现有新备份文件且晚于当前插件配置时间时自动获取最新配置；
3. 支持手动备份或还原配置数据

### Which area(s) would this feature affect? (Select all that apply)

Browser Extension

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sedationh**
- [x] ✨ New feature (feat)
- [ ] 🐛 Bug fix (fix)
- [ ] 📝 Documentation change (docs)
- [ ] 💄 UI/style change (style)
- [ ] ♻️ Code refactoring (refactor)
- [ ] ⚡ Performance improvement (perf)
- [ ] ✅ Test related (test)
- [ ] 🔧 Build or dependencies update (build)
- [ ] 🔄 CI/CD related (ci)
- [ ] 🌐 Internationalization (i18n)
- [ ] 🧠 AI model related (ai)
- [ ] 🔄 Revert a previous commit (revert)
- [ ] 📦 Other changes that do not modify src or test files (chore)

## Related Issue

Closes #286

## How Has This Been Tested?

- [ ] Added unit tests
- [ ] Verified through manual testing

## Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly if necessary
- [ ] My code follows the code style of this project
- [ ] My changes do not break existing functionality
- [ ] If my code was generated by AI, I have proofread and improved it as necessary.

## Summary by Sourcery

Add a config synchronization interface to the extension that enables users to import, export (with optional API key exclusion), view, migrate, and validate their configuration settings, along with necessary utilities and localization.

New Features:
- Introduce ConfigSync component with import, export, and view functionality for extension settings
- Support configuration migration on import to ensure compatibility with schema version changes
- Provide option to include or exclude API keys during export to protect sensitive information

Enhancements:
- Add getConfigWithoutAPIKeys and hasAPIKey utilities to strip or detect API keys in configurations
- Display a refresh icon on the reset config button
- Rename storedCSchemaVersion to storedConfigSchemaVersion for clearer schema version handling

Documentation:
- Add i18n translation strings for config sync UI in English, Japanese, Korean, Simplified Chinese, and Traditional Chinese

Chores:
- Add a changeset for bumping the extension version

### CODE REVIEW

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+import { configAtom, writeConfigAtom } from '@/utils/atoms/config'
+import { getConfigWithoutAPIKeys } from '@/utils/config/config'
+import { runMigration } from '@/utils/config/migration'
+import { APP_NAME } from '@/utils/constants/app'
+import { CONFIG_SCHEMA_VERSION } from '@/utils/constants/config'
+import { logger } from '@/utils/logger'
+import { ConfigCard } from '../../components/config-card'
+import { storedConfigSchemaVersionAtom } from './atom'
+
+function ConfigSync() {
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: You can use the `export default` function directly here if you are only exporting one component. Doing so allows others to immediately understand the primary purpose of this file upon opening it.

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  useEffect(() => {
+    (async () => {
+      const storedConfigSchemaVersion = await storage.getItem<number>(`local:__configSchemaVersion`)
+      setStoredConfigSchemaVersion(storedConfigSchemaVersion ?? 1)
+    })()
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  return (
+    <div className="space-y-6">
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: ConfigCard should give enough and equal padding for all sections? If you really feel need more space, should we add this class to ConfigCard to override the space instead?

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            // migrate
+            let currentVersion = importStoredConfigSchemaVersion
+            while (currentVersion < CONFIG_SCHEMA_VERSION) {
+              const nextVersion = currentVersion + 1
+              config = await runMigration(nextVersion, config)
+              currentVersion = nextVersion
+            }
+          }
+          if (!configSchema.safeParse(config).success) {
+            toast.error(i18n.t('options.config.sync.validationError'))
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx:82]: maybe can add the specific error from zod to the toast as well?

Comment (ananaBMaster) [reply apps/extension/src/entrypoints/options/pages/config/config-sync.tsx:82]: btw, have you tested the error ui as well?

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            logger.error(config, configSchema.safeParse(config).error)
+            return
+          }
+          setConfig(config)
+          toast.success(`${i18n.t('options.config.sync.importSuccess')} !`)
+        }
+        catch (error) {
+          logger.error(error)
+          toast.error(i18n.t('options.config.sync.importError'))
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx:91]: if the error has specific message, may should include it into the toast

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        }
+      }
+      reader.readAsText(file)
+    }
+    catch {
+      toast.error(i18n.t('options.config.sync.importError'))
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: if you gonna catch the error here finally, is it really necessary to catch and toast the importError before here?

```
catch (error) {
          logger.error(error)
          toast.error(i18n.t('options.config.sync.importError'))
```

the previous one seems to be duplicated catch for importError

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+      <AlertDialogContent>
+        <AlertDialogHeader>
+          <AlertDialogTitle>{i18n.t('options.config.sync.exportOptions.title')}</AlertDialogTitle>
+          <AlertDialogDescription>
+            {i18n.t('options.config.sync.exportOptions.description')}
+          </AlertDialogDescription>
+        </AlertDialogHeader>
+
+        <AlertDialogFooter className="flex !justify-between !flex-row">
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: can I ask why need `!` here?

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  )
+}
+
+function ViewCurrentConfig() {
+  const config = useAtomValue(configAtom)
+  const storedConfigSchemaVersion = useAtomValue(storedConfigSchemaVersionAtom)
+  const [isExpanded, setIsExpanded] = useState(false)
+
+  return (
+    <div className="w-full flex flex-col   justify-end">
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: too many spaces between `flex-col` and `justify-end`

File: apps/extension/src/utils/config/config.ts [hunk 8]
```diff
@@ -36,3 +36,56 @@ export function isAnyAPIKey(providersConfig: ProvidersConfig) {
...
+export function getConfigWithoutAPIKeys<T extends Record<string, any>>(config: T): T {
+  function deepClean(obj: any): any {
+    if (Array.isArray(obj)) {
+      return obj.map(deepClean)
+    }
+    if (obj && typeof obj === 'object') {
+      const newObj: Record<string, any> = {}
+      for (const key in obj) {
+        if (key === 'apiKey') {
+          continue
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/utils/config/config.ts:50]: @sedationh do you know what does this mean?

File: apps/extension/src/utils/config/config.ts [hunk 9]
```diff
@@ -36,3 +36,56 @@ export function isAnyAPIKey(providersConfig: ProvidersConfig) {
...
+        }
+        newObj[key] = deepClean(obj[key])
+      }
+      return newObj
+    }
+    return obj
+  }
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/utils/config/config.ts:57]: may need to deal with this? is there any warning in the local ide

**Review by ananaBMaster — CHANGES_REQUESTED** [review]

File: apps/extension/src/utils/config/__tests__/config.test.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { describe, expect, it } from 'vitest'
+import { getConfigWithoutAPIKeys, hasAPIKey } from '../config'
+import { LATEST_SCHEMA_VERSION } from '../migration'
+
+describe('config utilities', () => {
+  describe('getConfigWithoutAPIKeys', async () => {
+    for (let version = 2; version <= LATEST_SCHEMA_VERSION; version++) {
+      const currentVersionStr = String(version).padStart(3, '0')
+
+      it(`should remove api keys from config v${currentVersionStr}`, async () => {
```
Comment (ananaBMaster) [inline apps/extension/src/utils/config/__tests__/config.test.ts:10]: can add more tests. test success, test failure and corner cases. you can use ai to generate more

**Review by ananaBMaster — CHANGES_REQUESTED** [review]

File: apps/extension/src/entrypoints/options/pages/config/reset-config.tsx [hunk 11]
```diff
@@ -25,9 +26,14 @@ export function ResetConfig() {
   return (
     <ConfigCard title={i18n.t('options.config.resetConfig.title')} description={i18n.t('options.config.resetConfig.description')}>
       <AlertDialog>
-        <AlertDialogTrigger className="w-full flex items-center lg:justify-end">
-          <Button variant="destructive">{i18n.t('options.config.resetConfig.dialog.trigger')}</Button>
-        </AlertDialogTrigger>
+        <div className="w-full flex justify-end">
+          <AlertDialogTrigger>
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/entrypoints/options/pages/config/reset-config.tsx]: @sedationh 这个说的是对的

File: apps/extension/src/utils/config/__tests__/config.test.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import { describe, expect, it } from 'vitest'
+import { getConfigWithoutAPIKeys, hasAPIKey } from '../config'
+import { LATEST_SCHEMA_VERSION } from '../migration'
+
+describe('config utilities', () => {
+  describe('getConfigWithoutAPIKeys', async () => {
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/utils/config/__tests__/config.test.ts]: @sedationh 这个看过吗？

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+          const { storedConfigSchemaVersion: importStoredConfigSchemaVersion } = _config
+          if (importStoredConfigSchemaVersion > storedConfigSchemaVersion) {
+            toast.error(i18n.t('options.config.sync.versionTooNew'))
+            return
+          }
+          let { config } = _config
+          if (importStoredConfigSchemaVersion < storedConfigSchemaVersion) {
+            // migrate
+            let currentVersion = importStoredConfigSchemaVersion
+            while (currentVersion < CONFIG_SCHEMA_VERSION) {
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx:75]: 逻辑不太对，前面说 if (importStoredConfigSchemaVersion < storedConfigSchemaVersion) 那就进入下面的 migrate，但是 migrate，又想要 migrate 到 CONFIG_SCHEMA_VERSION，所以逻辑前后没对上。其实你这边是想要 migrate 到最新的 config version，那么前面的 if 判断是不是就应该是 importStoredConfigSchemaVersion < CONFIG_SCHEMA_VERSION

另外一点就是           
```
    if (importStoredConfigSchemaVersion > storedConfigSchemaVersion) {
            toast.error(i18n.t('options.config.sync.versionTooNew'))
            return
          }
```
这个有必要吗？ 我是不是只要判断 if (importStoredConfigSchemaVersion > CONFIG_SCHEMA_VERSION) 然后报错，就好了

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+import { logger } from '@/utils/logger'
+import { ConfigCard } from '../../components/config-card'
+import { storedConfigSchemaVersionAtom } from './atom'
+
+export default function ConfigSync() {
+  const setStoredConfigSchemaVersion = useSetAtom(storedConfigSchemaVersionAtom)
+
+  useEffect(() => {
+    (async () => {
+      const storedConfigSchemaVersion = await storage.getItem<number>(`local:__configSchemaVersion`)
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: 我仔细想了一下，感觉这个 storedConfigSchemaVersionAtom 在这里没有必要。可以完全移除整个 atom 在这个 PR

因为 `apps/extension/src/entrypoints/options/main.tsx` 里面，

```
async function initApp() {
  await loadGlobalConfig()
```
已经保证了 storage.local 里面的 `storedConfigSchemaVersion` 和 `CONFIG_SCHEMA_VERSION` 保持一致了。所以所有用到 storedConfigSchemaVersion 的地方，使用 `CONFIG_SCHEMA_VERSION` 即可。

如果维护一份 duplicated 的状态 `storedConfigSchemaVersion` 代码量变多，逻辑容易混乱，还要记得及时更新这份 duplicated 变量，也容易写出 bug。

所以所有地方`storedConfigSchemaVersion` 用 `CONFIG_SCHEMA_VERSION`  替代即可，并且不需要这个 useEffect

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+            return
+          }
+          const _config = JSON.parse(fileResult)
+          const { storedConfigSchemaVersion: importStoredConfigSchemaVersion } = _config
+          if (importStoredConfigSchemaVersion > storedConfigSchemaVersion) {
+            toast.error(i18n.t('options.config.sync.versionTooNew'))
+            return
+          }
+          let { config } = _config
+          if (importStoredConfigSchemaVersion < storedConfigSchemaVersion) {
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: 这个说的是对的，虽然它的原因说的不是很准确，可以结合我其他评论来看

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+        accept=".json"
+        onChange={importConfig}
+      />
+    </Button>
+  )
+}
+
+function ExportConfig() {
+  const config = useAtomValue(configAtom)
+  const storedConfigSchemaVersion = useAtomValue(storedConfigSchemaVersionAtom)
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: 结合上面的评论，这个也不需要了，直接用 `CONFIG_SCHEMA_VERSION`

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx (hunk 14 above)
Comment (ananaBMaster) [reply apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: 举个例子，你这个写法，会出现一个中间态。
当 initApp 执行完 await loadGlobalConfig()，然后后续 React 刚被 init，useEffect 的这个 async 函数并未完成的时候。你的 storage.local 里面已经是最新的 `local:__configSchemaVersion`，并且 `local:config` 也已经是最新，但是 atom 在并未完成 这个 await 的时候，storedConfigSchemaVersion 会不是最新的。这个中间态是错误的中间态，容易 bug

File: apps/extension/src/utils/config/__tests__/config.test.ts (hunk 10 above)
Comment (ananaBMaster) [reply apps/extension/src/utils/config/__tests__/config.test.ts:10]: 有时间的话，可以多加一些测试，没时间也没事。不过可以试试让 ai 加一些测试

**Review by ananaBMaster — CHANGES_REQUESTED** [review]
有一些 ai 说的其实是对的，所以我把一些 unresolve 了，然后我又加了些评论

File: apps/extension/src/utils/config/__tests__/config.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (sedationh) [reply apps/extension/src/utils/config/__tests__/config.test.ts]: get , i will delete async in describe

File: apps/extension/src/utils/config/__tests__/config.test.ts (hunk 10 above)
Comment (sedationh) [reply apps/extension/src/utils/config/__tests__/config.test.ts:10]: 这个没法加异常 case，搞的是全消除，群里里有

File: apps/extension/src/entrypoints/options/pages/config/config-sync.tsx (hunk 14 above)
Comment (sedationh) [reply apps/extension/src/entrypoints/options/pages/config/config-sync.tsx]: make sense
